### PR TITLE
Add error boundary to catch errors in visualizations

### PR DESCRIPTION
**Pull request by kravets-levko**
## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Bug Fix
- [x] Other

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

image
image

### CODE REVIEW

File: client/app/components/ErrorBoundary.jsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  renderError(error) {
+    if (isFunction(this.props.renderError)) {
+      return this.props.renderError(error);
+    }
+    return (
+      <Alert
+        message="Something went wrong."
+        description={
+          <Typography.Text style={{ display: "block" }} ellipsis>
+            {error.message}
```
Comment (arikfr) [inline client/app/components/ErrorBoundary.jsx]: Let's avoid showing users our dirty laundry and only log it (with `debug`). In the future it makes sense to send this into Sentry.

Comment (kravets-levko) [reply client/app/components/ErrorBoundary.jsx]: image

Comment (kravets-levko) [reply client/app/components/ErrorBoundary.jsx]: @arikfr Maybe, we could show some message instead - like "Please check visualization options"/"read docs"/link to knowledge base/"Your PC is tired, turn it off"/etc?

Comment (arikfr) [reply client/app/components/ErrorBoundary.jsx]: I would go with either one of:

1. `Error while rendering visualization. Please check visualization options and/or dataset.`
2. `Error while rendering visualization.`

I'm not sure about suggesting checking the option as usually an error boundary will be triggered due to a bug, no?

Comment (kravets-levko) [reply client/app/components/ErrorBoundary.jsx]: Yes, mostly that will be bugs. But I think there are situations when checking options may help - for example, when visualization expects some column to be, say, numeric, but it's date/time. Of course there should be sanity check in the visualization's code, but in this case user can just use another column or update query. Anyway, we could leave just a "Something went wrong" message

Comment (arikfr) [reply client/app/components/ErrorBoundary.jsx]: "Error while rendering visualization." is probably a bit better than the generic "something went wrong" as it gives a direction to what went wrong :)

Comment (kravets-levko) [reply client/app/components/ErrorBoundary.jsx]: Done https://github.com/getredash/redash/pull/4518/commits/7c5024b86438f51397ba30612c21fde0cf9e820f

image

File: client/app/visualizations/chart/Renderer/PlotlyChart.jsx [hunk 2]
```diff
@@ -1,42 +1,65 @@
...
 
 import getChartData from "../getChartData";
 import { Plotly, prepareData, prepareLayout, updateData, applyLayoutFixes } from "../plotly";
 
+function catchErrors(func, errorHandler) {
+  return (...args) => {
+    try {
+      return func(...args);
+    } catch (error) {
+      errorHandler.handleError(error);
```
Comment (gabrieldutra) [inline client/app/visualizations/chart/Renderer/PlotlyChart.jsx]: What is this part for? Overall it seemed it only needed a `<ErrorBoundary>` component as a fallback in the tree (no need to use context for example), so was there any issues or there are any gains when you add this part?

Comment (kravets-levko) [reply client/app/visualizations/chart/Renderer/PlotlyChart.jsx]: Error boundary will catch only errors that happened during rendering; this part is needed to catch async errors as well

**Comment by arikfr** [comment]
:+1:

**Pull request merged**